compiler/rmi/kotlinr/src/KotlinCompilerClient.kt

```diff
@@ -0,0 +1,262 @@
+            }
+            return null
+        }
+
+
+        private fun startDaemon(compilerId: CompilerId, daemonLaunchingOptions: DaemonLaunchingOptions, daemonOptions: DaemonOptions, errStream: PrintStream) {
+            val javaExecutable = listOf(System.getProperty("java.home"), "bin", "java").joinToString(File.separator)
+            // TODO: add some specific environment variables to the cp and may be command line, to allow some specific startup configs
+            val args = listOf(javaExecutable,
+                              "-cp", compilerId.compilerClasspath.joinToString(File.pathSeparator)) +
+                       daemonLaunchingOptions.jvmParams +
+                       COMPILER_DAEMON_CLASS_FQN +
+                       daemonOptions.asParams +
+                       compilerId.asParams
+            errStream.println("[daemon client] starting the daemon as: " + args.joinToString(" "))
+            val processBuilder = ProcessBuilder(args).redirectErrorStream(true)
+            // assuming daemon process is deaf and (mostly) silent, so do not handle streams
+            val daemon = processBuilder.start()
+
+            val lock = ReentrantReadWriteLock()
+            var isEchoRead = false
+
+            val stdouThread =
+                    thread {
+                        daemon.getInputStream()
+                              .reader()
+                              .forEachLine {
+                                  if (daemonOptions.startEcho.isNotEmpty() && it.contains(daemonOptions.startEcho))
+                                      lock.write { isEchoRead = true; return@forEachLine }
+                                  errStream.println("[daemon] " + it)
+                              }
+                    }
+            try {
+                // trying to wait for process
+                if (daemonOptions.startEcho.isNotEmpty()) {
+                    errStream.println("[daemon client] waiting for daemon to respond")
+                    var waitMillis: Long = DAEMON_STARTUP_TIMEOUT_MS / DAEMON_STARTUP_CHECK_INTERVAL_MS
+                    while (waitMillis-- > 0) {
+                        Thread.sleep(DAEMON_STARTUP_CHECK_INTERVAL_MS)
+                        if (!daemon.isAlive() || lock.read { isEchoRead } == true) break;
+                    }
+                    if (!daemon.isAlive())
+                        throw Exception("Daemon terminated unexpectedly")
+                    if (lock.read { isEchoRead } == false)
+                        throw Exception("Unable to get response from daemon in $DAEMON_STARTUP_TIMEOUT_MS ms")
+                }
+                else
+                // without startEcho defined waiting for max timeout
+                    Thread.sleep(DAEMON_STARTUP_TIMEOUT_MS)
+            }
+            finally {
+                // assuming that all important output is already done, the rest should be routed to the log by the daemon itself
+                if (stdouThread.isAlive)
```

Comment:
But if it is not then the child process could hang if there are bytes were not consumed: generally it may happen if the child process is trying to write to the pipe but nobody reads it anymore and the pipe internal buffer is full